hyperspy/io_plugins/usid_hdf5.py

```diff
@@ -0,0 +1,447 @@
+
+    if is_compound:
+        sig = []
+        # Iterate over each dimension name:
+        for name in ds_nd.dtype.names:
+            q_sub, u_sub = __split_descriptor(name)
+            # TODO: Check to make sure that this will work with Dask.array
+            sig.append(trunc_func(ds_nd[name], q_sub, u_sub, sig_type=quant))
+    else:
+        sig = [trunc_func(ds_nd, quant, units)]
+
+    return sig
+
+
+# ######################## UTILITIES THAT SIMPLIFY WRITING TO H5USID FILES #############################################
+
+
+def __flatten_nested_dictionary(nested_dict, parent_key='', sep='-'):
+    """
+    Flattens a nested dictionary
+
+    Parameters
+    ----------
+    nested_dict : dict
+        Nested dictionary
+    parent_key : str, Optional
+        Name of current parent
+    sep : str, Optional. Default='-'
+        Separator between the keys of different levels
+
+    Returns
+    -------
+    dict
+        Dictionary whose keys are flattened to a single level
+    Notes
+    -----
+    Taken from https://stackoverflow.com/questions/6027558/flatten-nested-dictionaries-compressing-keys
+    """
+    items = []
+    for k, v in nested_dict.items():
+        new_key = parent_key + sep + k if parent_key else k
+        if isinstance(v, collections.MutableMapping):
+            items.extend(__flatten_nested_dictionary(v, new_key, sep=sep).items())
+        else:
+            items.append((new_key, v))
+    return dict(items)
+
+
+def __axes_list_to_dimensions(axes_list, data_shape, is_spec):
+    dim_list = []
+    dim_type = 'Pos'
+    if is_spec:
+        dim_type = 'Spec'
+    # for dim_ind, (dim_size, dim) in enumerate(zip(data_shape, axes_list)):
+    for dim_ind in range(len(data_shape)):
+        dim_size = data_shape[len(data_shape) - 1 - dim_ind]
+        dim = axes_list[dim_ind]
+        dim_name = dim_type + '_Dim_' + str(dim_ind)
+        if isinstance(dim.name, str):
+            temp = dim.name.strip()
+            if len(temp) > 0:
+                dim_name = temp
+        dim_units = 'a. u.'
+        if isinstance(dim.units, str):
+            temp = dim.units.strip()
+            if len(temp) > 0:
+                dim_units = temp
+                # use REAL dimension size rather than what is presented in the axes manager
+        # print(dim_name, dim.size, dim_size) # dim_units, dim.offset, dim.scale,
+        dim_list.append(usid.Dimension(dim_name, dim_units,
+                                       np.arange(dim.offset,
+                                                 dim.offset + dim_size * dim.scale,
+                                                 dim.scale)))
+    if len(dim_list) == 0:
+        return usid.Dimension('Arb', 'a. u.', 1)
+    return dim_list[::-1]
+
+# ######################################################################################################################
+
+
+def file_reader(filename, dset_path='', verbose=False, ignore_non_linear_dims=True, **kwds):
```

Comment:
Set `dset_path` default to None? Usually, we load most of the data as a list of signals.

Add parameters specific to this reader to the `load` function: https://github.com/hyperspy/hyperspy/blob/145003e38598d8244625390bc3de368de7db985d/hyperspy/io.py#L46

Reply:
I don't think it would be a good idea to load dozens of datasets into memory (numpy). We can change the default value to None once lazy reading is implemented. Until then, I would recommend keeping the default to reading a single dataset.

Reply:
Now, ``dset_path=None`` is the default given that lazy loading is allowed.